### PR TITLE
feat(consensus): add withdraw helpers for all/percent/value

### LINKED ISSUE #462: Add convenience withdrawal helpers on top of shares-only withdrawStake
### Summary

Restore the previous developer experience for common withdrawal flows (all, percent, target stake) by adding a thin wrapper layer on top of the protocol-aligned `withdrawStake({ api, operatorId, shares })`. Leave `withdrawStake` unchanged.

### Background

- The SDK was updated to align with protocol changes requiring shares-only withdrawal: see [PR #460](https://github.com/autonomys/auto-sdk/pull/460).
- While correct at the protocol layer, this removed ergonomic options that many apps relied on.

### Problem

Developers commonly need to:

- Withdraw all stake for an account on an operator
- Withdraw by percent
- Withdraw an approximate balance amount (in tokens)

These now require callers to compute shares off-chain, increasing boilerplate and potential for mistakes.

### Proposal

Add three convenience helpers that compute shares and delegate to `withdrawStake`:

```ts
// Returns a submittable extrinsic
const withdrawStakeAll = (params: { api: Api, operatorId: StringNumberOrBigInt, account: string })

const withdrawStakeByPercent = (params: {
  api: Api
  operatorId: StringNumberOrBigInt
  account: string
  percent: StringNumberOrBigInt // 0..100, rounded down to shares
})

const withdrawStakeByValue = (params: {
  api: Api
  operatorId: StringNumberOrBigInt
  account: string
  amountToWithdraw: StringNumberOrBigInt // balance units
})
```
Implementation details:

- Use `nominatorPosition(api, operatorId, account)` to get `totalShares` (user’s current shares) and `currentStakedValue` (user’s current stake value at price).
- For percent: `shares = floor(totalShares * percent / 100)`, clamp to `[0, totalShares]`.
- For value: `shares = floor(min(requestedAmount, currentStakedValue) * totalShares / currentStakedValue)` using integers.
- Edge cases: return clear errors for zero shares, clamp percent, and handle `totalShares == 0`.

### Naming

- Keep `withdrawStake` as-is (shares-only, protocol-native).
- Add the following wrappers: `withdrawStakeAll`, `withdrawStakeByPercent`, `withdrawStakeByValue`.

### File placement

- New module: `packages/auto-consensus/src/staking.withdrawal.helpers.ts`
- Re-export from `packages/auto-consensus/src/index.ts`

### Acceptance Criteria

- `withdrawStakeAll`, `withdrawStakeByPercent`, `withdrawStakeByValue` implemented and exported.
- Unit tests cover share calculation, rounding, clamping, and error paths.
- Examples in package README updated to demonstrate usage.
- No changes to `withdrawStake` signature or behavior.

### Tasks

- [ ] Implement helper module and exports
- [ ] Add unit tests
- [ ] Update docs and examples

### References

- Protocol alignment: [PR #460](https://github.com/autonomys/auto-sdk/pull/460)

### PR DESCRIPTION
**Pull request by jfrank-summit**
This pull request introduces new staking withdrawal helper functions to the `@autonomys/auto-consensus` SDK, making it easier for developers to withdraw stake by all shares, by percent, or by target value. It also refactors internal imports for clarity and adds comprehensive tests for these new helpers. The documentation is updated to reflect these enhancements. Closes #462 

**Staking Withdrawal Helper Functions:**

* Added new helper functions: `withdrawStakeAll`, `withdrawStakeByPercent`, and `withdrawStakeByValue` in `src/staking/withdrawal-helpers.ts`. These helpers simplify withdrawing all, a percentage, or a target value of stake for a nominator, including input validation and edge case handling.
* Exported the new withdrawal helpers in `src/staking/index.ts` for external usage.

**Testing:**

* Added a new test suite `staking.withdrawal.helpers.test.ts` to cover all edge cases and expected behaviors for the withdrawal helpers, ensuring robust error handling and correct share calculations.

**Documentation Updates:**

* Updated the README to document the new withdrawal helper functions with usage examples and clarified the staking functionality description. [[1]](diffhunk://#diff-f3214e618699372e2f6cba4d3fb293957c9e60621c065cbc47b36bf4fb930304L25-R25) [[2]](diffhunk://#diff-f3214e618699372e2f6cba4d3fb293957c9e60621c065cbc47b36bf4fb930304R191-R227)

**Internal Refactoring:**

* Refactored imports to consistently reference staking logic from `src/staking/staking.ts` in test and implementation files for better code organization. [[1]](diffhunk://#diff-122a1e09839fb512e759c58e034ab6d2353106e1b30c2e41640d83631eb7a5e1L2-R3) [[2]](diffhunk://#diff-428f1699174ec3e5933e254412586696dc9cb2003a4076e6b111e58120196e6aL3-R3)

**Core Staking Logic:**

* Added the new file `src/staking/staking.ts` to encapsulate all core staking operations, including operator management and withdrawal transactions, with improved error handling.

### CODE REVIEW

File: packages/auto-consensus/src/staking/withdrawal-helpers.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  const { totalShares, currentStakedValue } = position
+
+  if (totalShares === BigInt(0)) throw new Error('No shares to withdraw for the given account')
+  if (currentStakedValue === BigInt(0))
+    throw new Error('Current staked value is zero; cannot compute shares')
+
+  const effectiveAmount =
+    requestedAmount > currentStakedValue ? currentStakedValue : requestedAmount
+
+  const shares = (effectiveAmount * totalShares) / currentStakedValue
```
[the remark this answers is not in the document]
Comment (jfrank-summit) [reply packages/auto-consensus/src/staking/withdrawal-helpers.ts]: I think this is a user perspective issue. I'm not sure if users expect to receive **at least** the amount or if (at point of calculation) they expect the amount they requested is a ceiling, i.e. don't withdraw **more than** I requested. I tend to think it is the latter. Of course the value per share may change from calculation time to settlement time but I think that is a separate issue.

**Comment by jfrank-summit** [comment]
Storage fund refund was neglected in the initial PR. I have updated the `withdrawStakeByValue` function to adjust share withdrawal amount to account for storage fund refund.

**Comment by jfrank-summit** [comment]
bugbot run

File: packages/auto-consensus/src/staking/withdrawal-helpers.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (teor2345) [reply packages/auto-consensus/src/staking/withdrawal-helpers.ts]: Whatever is least surprising to users seems best here.

If there's no clear convention, then taking out slightly less might be best. (Like we round down fractional values when getting paid money in other ways.) Users will have to over-estimate their share amounts to compensate for price changes anyway.

It's also relatively simple to recover from withdrawing fewer shares - just withdraw a few more. But adding more stake is slightly more complicated.
